pydantic_ai_slim/pydantic_ai/toolsets/run.py

```diff
@@ -0,0 +1,142 @@
+from __future__ import annotations
+
+from collections.abc import Iterable, Iterator
+from contextlib import contextmanager
+from dataclasses import dataclass, replace
+from typing import Any
+
+from pydantic import ValidationError
+
+from pydantic_ai.output import DeferredToolCalls
+
+from .. import messages as _messages
+from .._run_context import AgentDepsT, RunContext
+from ..exceptions import ModelRetry, ToolRetryError, UnexpectedModelBehavior
+from ..tools import ToolDefinition
+from . import AbstractToolset
+from .wrapper import WrapperToolset
+
+
+@dataclass(init=False)
+class RunToolset(WrapperToolset[AgentDepsT]):
+    """A toolset that is frozen for a specific run."""
```

Comment:
??? what is this?

Reply:
I was expecting this to raise some questions 😄 And the docstring is wrong because it's really frozen for each run _step_ instead of an entire run.

The idea is that at each run step, all the toolsets need to prepared (because the prepare functions take run context) and then frozen (meaning the tool defs are cached) so that any changes to the underlying toolsets don't take effect until the next run step (when the toolsets wrapped by this `RunToolset` are again prepared and re-wrapped). We also use this preparation stage to request the list of MCP tools (and cache them in the `RunToolset`) as we're not connected to the MCP server yet when the toolsets are initially built.